Assets/Mirage/Runtime/InterestManagement/InterestManager.cs

```diff
@@ -0,0 +1,286 @@
+        public InterestManager(ServerObjectManager serverObjectManager, int initialSystems = 0)
+        {
+            _initialSystems = initialSystems;
+
+            ServerObjectManager = serverObjectManager;
+
+            ServerObjectManager.Server.Started.AddListener(OnServerStarted);
+            ServerObjectManager.Server.Stopped.AddListener(OnServerStopped);
+        }
+
+        /// <summary>
+        /// Send a message to all observers of an identity
+        /// </summary>
+        /// <param name="identity"></param>
+        /// <param name="msg"></param>
+        /// <param name="channelId"></param>
+        protected internal virtual void Send<T>(NetworkIdentity identity, T msg, int channelId = Channel.Reliable, INetworkPlayer skip = null)
+        {
+            var stopWatch = Stopwatch.StartNew();
+
+            HashSet<INetworkPlayer> observers = Observers(identity);
+
+            if (observers.Count == 0)
+                return;
+
+            using (PooledNetworkWriter writer = NetworkWriterPool.GetWriter())
+            {
+                // pack message into byte[] once
+                MessagePacker.Pack(msg, writer);
+                var segment = writer.ToArraySegment();
+                int count = Send(identity, segment, channelId, skip);
+
+                if (count > 0)
+                    NetworkDiagnostics.OnSend(msg, segment.Count, count);
+            }
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - Send Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        /// Send a message to all observers of an identity
+        /// </summary>
+        /// <remarks>Override if you wish to provide
+        /// an allocation free send method</remarks>
+        /// <param name="identity">the object that wants to send a message</param>
+        /// <param name="data">the data to send</param>
+        /// <param name="channelId">the channel to send it on</param>
+        /// <param name="skip">a player who should not receive the message</param>
+        /// <returns>Total amounts of messages sent</returns>
+        protected virtual int Send(NetworkIdentity identity, ArraySegment<byte> data, int channelId = Channel.Reliable, INetworkPlayer skip = null)
+        {
+            int count = 0;
+
+            foreach (INetworkPlayer player in Observers(identity))
+            {
+                if (player == null) continue;
+
+                if (player != skip)
+                {
+                    // send to all connections, but don't wait for them
+                    player.Send(data, channelId);
+                    count++;
+                }
+            }
+
+            return count;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckGrowVisibilitySystem()
```

Comment:
It looks like this function just grows an array? wouldn't it be simpler to just use a list?

Reply:
Reason I use arrays is for speed. The array should only ever grow on first initialization. I set a small number of 4 at first because i am not expecting more then 4 visibility systems to ever be in play. So list would just be redundant i think for this.

Reply:
I think this is one of those optimization that requires to benchmark because a List is so much simpler